applications/crossbar/src/modules/cb_modules_util.erl

```diff
@@ -563,29 +564,132 @@ remove_plaintext_password(Context) ->
+    AssignResults = maybe_assign_to_app(NumUpdates, AccountId),
+    PortAssignResults ++ AssignResults.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Split a list of assignment updates into a 2-element tuple; element
+%% 1 is a list of port requests, element 2 is a list of numbers that
+%% are already active.
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec split_port_requests(assignment_to_apply(), {port_req_assignments(), assignments_to_apply()}) ->
+                                 {port_req_assignments(), assignments_to_apply()}.
+split_port_requests({DID, Assign}=ToApply, {PRUpdates, NumUpdates}) ->
     Num = knm_converters:normalize(DID),
     case knm_port_request:get(Num) of
-        {'error', _} ->
-            {DID, knm_number:assign_to_app(DID, Assign)};
         {'ok', JObj} ->
-            {DID, knm_port_request:assign_to_app(Num, Assign, JObj)}
+            {[{Num, Assign, JObj}|PRUpdates], NumUpdates};
+        {'error', _} ->
+            {PRUpdates, [ToApply|NumUpdates]}
     end.
 
+-spec assign_to_port_number(port_req_assignments()) ->
+                                   assignment_updates().
+assign_to_port_number(PRUpdates) ->
+    lists:foldl(fun({Num, Assign, JObj}, Acc) ->
+                        [{Num, knm_port_request:assign_to_app(Num, Assign, JObj)}
+                         | Acc]
+                end, [], PRUpdates).
+
+-spec maybe_assign_to_app(assignments_to_apply(), ne_binary()) ->
+                                 assignment_updates().
+maybe_assign_to_app(NumUpdates, AccountId) ->
+    Options = [{'auth_by', AccountId}],
+    Groups = group_by_assign_to(NumUpdates),
+    maps:fold(fun(Assign, Nums, Acc) ->
+                      Results = knm_numbers:assign_to_app(Nums, Assign, Options),
+                      format_assignment_results(Results) ++ Acc
+              end, [], Groups).
+
+-type assign_to_groups() :: #{api_binary() => ne_binaries()}.
+
+-spec group_by_assign_to(assignments_to_apply()) -> assign_to_groups().
+group_by_assign_to(NumUpdates) ->
+    group_by_assign_to(NumUpdates, #{}).
+
+group_by_assign_to([], Groups) -> Groups;
+group_by_assign_to([{DID, Assign}|NumUpdates], Groups) ->
+    DIDs = maps:get(Assign, Groups, []),
+    Groups1 = Groups#{Assign => [DID|DIDs]},
+    group_by_assign_to(NumUpdates, Groups1).
+
+-spec format_assignment_results(knm_numbers:ret()) -> assignment_updates().
+format_assignment_results(#{ok := OKs
+                           ,ko := KOs}) ->
+    format_assignment_oks(OKs) ++ format_assignment_kos(maps:to_list(KOs)).
+
+-spec format_assignment_oks(knm_number:knm_numbers()) -> assignment_updates().
+format_assignment_oks(Numbers) ->
+    lists:foldl(fun(Number, Acc) ->
+                        PN = knm_number:phone_number(Number),
+                        [{knm_phone_number:number(PN), {'ok', Number}} | Acc]
+                end, [], Numbers).
```

Comment:
@fenollp I have question about a list comprehension in this case.
Notice how I am assigning `PN = knm_number:phone_number(Number)`.
I can't seem to find that there is a list comprehension syntax to replicate that. I would be looking for something like:
`[{knm_phone_number(PN), {'ok', Number}} || Number <- Numbers, PN = knm_number:phone_number(Number)].`
Does such a thing not exist? It's not hugely necessary here, because I could do
`[{knm_phone_number:number(knm_number:phone_number(Number)), {'ok', Number}} || Number <- Numbers].`
But I could see that getting really clunky if I wanted to use the PN many times in the left side of my list comprehension.